workflow-manager/main.go

```diff
@@ -53,16 +53,106 @@ type batch struct {
+	pathComponents := strings.Split(batchName, "/")
+	batchID := pathComponents[len(pathComponents)-1]
+	aggregationID := pathComponents[0]
+	batchDate := pathComponents[1 : len(pathComponents)-1]
+
+	if len(batchDate) != 5 {
+		return batchPath{}, fmt.Errorf("malformed date in %q. Expected 5 date components, got %d", batchName, len(batchDate))
+	}
+
+	return batchPath{
+		aggregationID:       aggregationID,
+		batchDateComponents: batchDate,
+		batchID:             batchID,
+	}, nil
+}
+
+func (b *batchPath) batchPath() string {
+	return strings.Join([]string{b.aggregationID, b.batchDateString(), b.batchID}, "/")
+}
+
+func (b *batchPath) batchDateString() string {
+	return strings.Join(b.batchDateComponents, "/")
+}
+
+func (b *batchPath) batchDateTime() (time.Time, error) {
+	var dateComponents []int
+	for _, c := range b.batchDateComponents {
+		parsed, err := strconv.ParseInt(c, 10, 64)
+		if err != nil {
+			return time.Time{}, fmt.Errorf("parsing date component %q in %q: %w", c, b.batchPath(), err)
+		}
+		dateComponents = append(dateComponents, int(parsed))
+	}
+	return time.Date(dateComponents[0], time.Month(dateComponents[1]),
+		dateComponents[2], dateComponents[3], dateComponents[4], 0, 0, time.UTC), nil
+}
+
+// investionBasename returns the full name of the batch with any of the type
+// suffixes removed. All files that are part of a batch have the same basename.
+func ingestionBasename(s string) string {
 	s = strings.TrimSuffix(s, ".batch")
 	s = strings.TrimSuffix(s, ".batch.avro")
 	s = strings.TrimSuffix(s, ".batch.sig")
 	return s
 }
 
+// validationBasename returns the full name of the batch with any of the type
+// suffixes removed. All files that are part of a batch have the same basename.
+func validationBasename(s string, isFirst bool) string {
+	s = strings.TrimSuffix(s, fmt.Sprintf(".validity_%s", index(isFirst)))
+	s = strings.TrimSuffix(s, fmt.Sprintf(".validity_%s.avro", index(isFirst)))
+	s = strings.TrimSuffix(s, fmt.Sprintf(".validity_%s.sig", index(isFirst)))
+	return s
+}
+
+func parseIsFirst(isFirst string) bool {
+	return isFirst == "true"
+}
+
+// index returns the appropriate string to use in construction of filenames
+// based on whether the file creator was the "first" aka PHA server.
+func index(isFirst bool) string {
+	if isFirst {
+		return "0"
+	} else {
+		return "1"
+	}
+}
+
+func contains(s []string, str string) bool {
+	for _, v := range s {
+		if v == str {
+			return true
+		}
+	}
+
+	return false
+}
+
+func checkStoragePath(arg string, val string) (string, string) {
+	if val == "" {
+		log.Fatalf("%s is requored", arg)
+	}
+	if !strings.HasPrefix(val, "s3://") && !strings.HasPrefix(val, "gs://") {
+		log.Fatalf("invalid %s: %q", arg, val)
+	}
+
+	return val[0:2], val[4:]
+}
+
 var k8sNS = flag.String("k8s-namespace", "", "Kubernetes namespace")
+var isFirst = flag.String("is-first", "false", "Whether this set of servers is \"first\", aka PHA servers")
```

Comment:
I'd recommend a bool flag instead. Then you can get rid of parseIsFirst. https://golang.org/pkg/flag/#Bool